### PR TITLE
Fix bug in `QuasiAdamsBashforth2TimeStepper`

### LINKED ISSUE #2259: Euler time-stepping in the `QuasiAdamsBashforth2TimeStepper` is buggy --- a NaN will persist!
The Euler time stepping here is done by setting `χ=-0.5`

https://github.com/CliMA/Oceananigans.jl/blob/383e5b60ca8d3bfb3c8497b11b6be30946a25728/src/TimeSteppers/quasi_adams_bashforth_2.jl#L55

which then at 

https://github.com/CliMA/Oceananigans.jl/blob/383e5b60ca8d3bfb3c8497b11b6be30946a25728/src/TimeSteppers/quasi_adams_bashforth_2.jl#L138

implies that `G⁻` is multiplied with zeros (and thus ignored!).

But, what if `G⁻` includes a NaN? since `0 * NaN = NaN` then we don't get the expected behaviour here!

This (heroic) discovery was made by @glwagner after a week (or weeks) of trying to understand why NaN particles persisted in http://github.com/CliMA/OceanTurbulenceParameterEstimation.jl. See also http://github.com/CliMA/OceanTurbulenceParameterEstimation.jl/pulls/210.

### PR DESCRIPTION
**Pull request by navidcy**
This PR is an attempt/discussion on how to resolve #2259.

### CODE REVIEW

**Comment by glwagner** [comment]
A few ideas:

1) If we're doing an euler step, zero out the `G⁻` tendencies beforehand.

2) Put an if statement in to take `euler_step!` rather than `ab2_step!`.

The problem with 2) is that we haven't figured out how to abstract the stepping interface to handle different models. Basically, we don't really have an generic PDE time-stepper. Instead we have manually implemented time-stepping schemes for the different models; for example:

https://github.com/CliMA/Oceananigans.jl/blob/383e5b60ca8d3bfb3c8497b11b6be30946a25728/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl#L14-L36

Thus to achieve 2) we'll probably have to write an `euler_step!` for the hydrostatic free surface model. Maybe not a problem?

File: src/TimeSteppers/quasi_adams_bashforth_2.jl [hunk 1]
```diff
@@ -53,9 +53,19 @@ function time_step!(model::AbstractModel{<:QuasiAdamsBashforth2TimeStepper}, Δt
...
     χ = ifelse(euler, convert(eltype(model.grid), -0.5), model.timestepper.χ)
 
-    euler && @debug "Taking a forward Euler step."
+    if euler
+        @debug "Taking a forward Euler step."
+
+        # make sure we zero out all previous tendency fields to avoind errors
+        # in case G⁻ includes NaNs. See https://github.com/CliMA/Oceananigans.jl/issues/2259
+        for fieldname in keys(model.timestepper.G⁻)
+            field = model.timestepper.G⁻[fieldname]
```
Comment (glwagner) [inline src/TimeSteppers/quasi_adams_bashforth_2.jl:65]:
```suggestion
        for field in model.timestepper.G⁻
```

File: src/TimeSteppers/quasi_adams_bashforth_2.jl [hunk 2]
```diff
@@ -53,9 +53,19 @@ function time_step!(model::AbstractModel{<:QuasiAdamsBashforth2TimeStepper}, Δt
...
+            if !isnothing(field); parent(field) .= 0; end
```
Comment (glwagner) [inline src/TimeSteppers/quasi_adams_bashforth_2.jl:66]:
```suggestion
            !isnothing(field) && fill!(parent(field), 0)
```

**Review by glwagner — APPROVED** [review]

**Comment by navidcy** [comment]
Now a test ;)

**Comment by navidcy** [comment]
Hm...

```Julia
julia> model = HydrostaticFreeSurfaceModel(grid=RectilinearGrid(arch, size=(1, 1, 1), extent=(1, 2, 3)),
                                           buoyancy = BuoyancyTracer(),
                                           tracers = (:b, :e))
HydrostaticFreeSurfaceModel{CPU, Float64}(time = 0 seconds, iteration = 0)
├── grid: 1×1×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 1×1×1 halo
├── tracers: (:b, :e)
├── closure: Nothing
├── buoyancy: Buoyancy{BuoyancyTracer, Oceananigans.Grids.ZDirection}
├── free surface: ExplicitFreeSurface with gravitational acceleration 9.80665 m s⁻²
└── coriolis: Nothing

julia> keys(model.timestepper.G⁻)
(:u, :v, :η, :b, :e)
```

So it seems like `w` is not included. So perhaps the `!isnothing(field)` is not necessary?

**Comment by navidcy** [comment]
On the other hand checking for `!isnothing(field)` doesn't hurt...

**Comment by glwagner** [comment]
Ah, forget it then! It's a nice "implicit" check that the timestepper tendency makes sense.

**Comment by navidcy** [comment]
So, on the `main` branch I do this:

```Julia
julia> model = HydrostaticFreeSurfaceModel(grid=RectilinearGrid(size=(1, 1, 1), extent=(1, 2, 3)),
                                           buoyancy = BuoyancyTracer(),
                                           tracers = (:b, :e))
HydrostaticFreeSurfaceModel{CPU, Float64}(time = 0 seconds, iteration = 0)
├── grid: 1×1×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 1×1×1 halo
├── tracers: (:b, :e)
├── closure: Nothing
├── buoyancy: Buoyancy{BuoyancyTracer, Oceananigans.Grids.ZDirection}
├── free surface: ExplicitFreeSurface with gravitational acceleration 9.80665 m s⁻²
└── coriolis: Nothing

julia> model.timestepper.G⁻.η[1, 1, 1]
0.0

julia> model.timestepper.G⁻.η[1, 1, 1] = NaN
NaN

julia> model.timestepper.G⁻.η[1, 1, 1] * 0
NaN

julia> time_step!(model, 1)

julia> model.timestepper.G⁻.η[1, 1, 1]
0.0

julia> model.timestepper.Gⁿ.η[1, 1, 1]
0.0
```

Which makes me wonder: why don't I see NaNs?!

**Comment by navidcy** [comment]
On **`main`** branch:

```julia
julia> using Oceananigans

julia> model = HydrostaticFreeSurfaceModel(grid=RectilinearGrid(size=(1, 1, 1), extent=(1, 2, 3)),
                                           buoyancy = BuoyancyTracer(),
                                           tracers = (:b, :e))
HydrostaticFreeSurfaceModel{CPU, Float64}(time = 0 seconds, iteration = 0)
├── grid: 1×1×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 1×1×1 halo
├── tracers: (:b, :e)
├── closure: Nothing
├── buoyancy: Buoyancy{BuoyancyTracer, Oceananigans.Grids.ZDirection}
├── free surface: ExplicitFreeSurface with gravitational acceleration 9.80665 m s⁻²
└── coriolis: Nothing

julia> model.timestepper.Gⁿ.η[1, 1, 1]
0.0

julia> model.timestepper.G⁻.η[1, 1, 1]
0.0

julia> model.timestepper.G⁻.η[1, 1, 1] = NaN
NaN

julia> time_step!(model, 1; euler=true)

julia> model.timestepper.G⁻.η[1, 1, 1]
0.0

julia> model.timestepper.Gⁿ.η[1, 1, 1]
0.0
```

Why NaNs don't persist?

**Comment by navidcy** [comment]
Actually, on `**main**` the NaNs do appear, but after a bit more time-stepping:

```Julia
julia> model = HydrostaticFreeSurfaceModel(grid=RectilinearGrid(size=(1, 1, 1), extent=(1, 2, 3)),
                                           buoyancy = BuoyancyTracer(),
                                           tracers = (:b, :e))
HydrostaticFreeSurfaceModel{CPU, Float64}(time = 0 seconds, iteration = 0)
├── grid: 1×1×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 1×1×1 halo
├── tracers: (:b, :e)
├── closure: Nothing
├── buoyancy: Buoyancy{BuoyancyTracer, Oceananigans.Grids.ZDirection}
├── free surface: ExplicitFreeSurface with gravitational acceleration 9.80665 m s⁻²
└── coriolis: Nothing

julia> model.timestepper.G⁻.η[1, 1, 1] = NaN
NaN

julia> time_step!(model, 1e-9)

julia> model.timestepper.G⁻.η[1, 1, 1]
0.0

julia> model.timestepper.Gⁿ.η[1, 1, 1]
0.0

julia> time_step!(model, 1e-9)

julia> model.timestepper.G⁻.η[1, 1, 1]
0.0

julia> model.timestepper.Gⁿ.η[1, 1, 1]
0.0

julia> time_step!(model, 1e-9)

julia> model.timestepper.G⁻.η[1, 1, 1]
NaN

julia> model.timestepper.Gⁿ.η[1, 1, 1]
NaN

julia> model.timestepper.Gⁿ.η[1, 1, 1]
NaN

julia> time_step!(model, 1e-9)

julia> model.timestepper.G⁻.η[1, 1, 1]
NaN

julia> model.timestepper.Gⁿ.η[1, 1, 1]
NaN
```

**Comment by glwagner** [comment]
I would try `u` because the behavior of eta could depend on whether you use explicit or implicit time-stepping.

**Comment by navidcy** [comment]
```Julia
julia> model = HydrostaticFreeSurfaceModel(grid=RectilinearGrid(size=(1, 1, 1), extent=(1, 2, 3)),
                                           buoyancy = BuoyancyTracer(),
                                           tracers = (:b, :e))
HydrostaticFreeSurfaceModel{CPU, Float64}(time = 0 seconds, iteration = 0)
├── grid: 1×1×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 1×1×1 halo
├── tracers: (:b, :e)
├── closure: Nothing
├── buoyancy: Buoyancy{BuoyancyTracer, Oceananigans.Grids.ZDirection}
├── free surface: ExplicitFreeSurface with gravitational acceleration 9.80665 m s⁻²
└── coriolis: Nothing

julia> model.timestepper.G⁻.u[1, 1, 1] = NaN
NaN

julia> time_step!(model, 1e-9)

julia> model.timestepper.G⁻.u[1, 1, 1]
0.0

julia> model.timestepper.Gⁿ.u[1, 1, 1]
0.0

julia> time_step!(model, 1e-9)

julia> model.timestepper.G⁻.u[1, 1, 1]
NaN

julia> model.timestepper.Gⁿ.u[1, 1, 1]
NaN
```

**Comment by glwagner** [comment]
When you put a NaN in `G-.u`, does a NaN appear in `model.velocities.u` after a time-step?

**Comment by navidcy** [comment]
Good point! Yes.

OK. Test is being written!

**Comment by navidcy** [comment]
@glwagner have a look!

File: src/TimeSteppers/quasi_adams_bashforth_2.jl [hunk 3]
```diff
@@ -52,9 +53,17 @@ function time_step!(model::AbstractModel{<:QuasiAdamsBashforth2TimeStepper}, Δt
...
+    
     χ = ifelse(euler, convert(eltype(model.grid), -0.5), model.timestepper.χ)
 
-    euler && @debug "Taking a forward Euler step."
+    if euler
+        @debug "Taking a forward Euler step."
+        # Ensure zeroing out all previous tendency fields to avoid errors in
+        # case G⁻ includes NaNs. See https://github.com/CliMA/Oceananigans.jl/issues/2259
+        for field in model.timestepper.G⁻
+            !isnothing(field) && fill!(parent(field), 0)
```
Comment (glwagner) [inline src/TimeSteppers/quasi_adams_bashforth_2.jl:64]:
```suggestion
            fill!(parent(field), 0)
```

File: test/test_time_stepping.jl [hunk 4]
```diff
@@ -15,6 +15,17 @@ function time_stepping_works_with_flat_dimensions(arch, topology)
...
+function euler_time_stepping_doesnt_keep_NaNs(arch)
+    model = HydrostaticFreeSurfaceModel(grid=RectilinearGrid(size=(1, 1, 1), extent=(1, 2, 3)),
+                                        buoyancy = BuoyancyTracer())
+
+    model.timestepper.G⁻.u[1, 1, 1] = NaN
+
+    time_step!(model, 1, euler=true)
+
+    return !isnan.(model.velocities.u[1, 1, 1])
+end
```
Comment (glwagner) [inline test/test_time_stepping.jl:27]:
```suggestion
function euler_time_stepping_doesnt_keep_NaNs(arch)
    model = HydrostaticFreeSurfaceModel(grid=RectilinearGrid(arch, size=(1, 1, 1), extent=(1, 2, 3)),
                                        buoyancy = BuoyancyTracer())

    model.timestepper.G⁻.u[1, 1, 1] = NaN
    time_step!(model, 1, euler=true)
    u111 = CUDA.@allowscalar model.velocities.u[1, 1, 1]
    return !isnan(u111)
end
```

File: test/test_time_stepping.jl [hunk 5]
```diff
@@ -286,6 +297,13 @@ timesteppers = (:QuasiAdamsBashforth2, :RungeKutta3)
         end
     end
 
+    @testset "Euler time stepping propagate NaNs in previous tendency G⁻" begin
+        for arch in archs
+            @info "  Testing that Euler time stepping doesn't propagate NaNs found in previous tendency G⁻ [$(typeof(arch))]..."
+            @test euler_time_stepping_doesnt_keep_NaNs(arch)
```
Comment (glwagner) [inline test/test_time_stepping.jl:302]:
```suggestion
            @test euler_time_stepping_doesnt_propagate_NaNs(arch)
```

File: test/test_time_stepping.jl [hunk 6]
```diff
@@ -15,6 +15,16 @@ function time_stepping_works_with_flat_dimensions(arch, topology)
     return true # Test that no errors/crashes happen when time stepping.
 end
 
+function euler_time_stepping_doesnt_keep_NaNs(arch)
```
Comment (glwagner) [inline test/test_time_stepping.jl:18]:
```suggestion
function euler_time_stepping_doesnt_propagate_NaNs(arch)
```

File: src/TimeSteppers/quasi_adams_bashforth_2.jl [hunk 7]
```diff
@@ -52,9 +53,17 @@ function time_step!(model::AbstractModel{<:QuasiAdamsBashforth2TimeStepper}, Δt
...
+    
     χ = ifelse(euler, convert(eltype(model.grid), -0.5), model.timestepper.χ)
 
-    euler && @debug "Taking a forward Euler step."
+    if euler
+        @debug "Taking a forward Euler step."
+        # Ensure zeroing out all previous tendency fields to avoid errors in
+        # case G⁻ includes NaNs. See https://github.com/CliMA/Oceananigans.jl/issues/2259
+        for field in model.timestepper.G⁻
+            fill!(parent(field), 0)
```
Comment (glwagner) [inline src/TimeSteppers/quasi_adams_bashforth_2.jl:64]:
```suggestion
            field .= 0
```

I guess the code I suggested doesn't work on the cubed sphere. We can try the above! We might need to fix broadcasting but I can do that.

File: src/TimeSteppers/quasi_adams_bashforth_2.jl [hunk 8]
```diff
@@ -52,9 +53,17 @@ function time_step!(model::AbstractModel{<:QuasiAdamsBashforth2TimeStepper}, Δt
...
+    
     χ = ifelse(euler, convert(eltype(model.grid), -0.5), model.timestepper.χ)
 
-    euler && @debug "Taking a forward Euler step."
+    if euler
+        @debug "Taking a forward Euler step."
+        # Ensure zeroing out all previous tendency fields to avoid errors in
+        # case G⁻ includes NaNs. See https://github.com/CliMA/Oceananigans.jl/issues/2259
+        for field in model.timestepper.G⁻
+            field .= 0
```
Comment (glwagner) [inline src/TimeSteppers/quasi_adams_bashforth_2.jl:64]:
```suggestion
            !isnothing(field) && (field .= 0)
```

File: src/TimeSteppers/quasi_adams_bashforth_2.jl [hunk 9]
```diff
@@ -52,9 +53,17 @@ function time_step!(model::AbstractModel{<:QuasiAdamsBashforth2TimeStepper}, Δt
...
+    
     χ = ifelse(euler, convert(eltype(model.grid), -0.5), model.timestepper.χ)
 
-    euler && @debug "Taking a forward Euler step."
+    if euler
+        @debug "Taking a forward Euler step."
+        # Ensure zeroing out all previous tendency fields to avoid errors in
+        # case G⁻ includes NaNs. See https://github.com/CliMA/Oceananigans.jl/issues/2259
+        for field in model.timestepper.G⁻
+            !isnothing(field) && (field .= 0)
```
Comment (glwagner) [inline src/TimeSteppers/quasi_adams_bashforth_2.jl:64]:
```suggestion
            !isnothing(field) && fill!(field, 0)
```

**Pull request merged**